core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -106,7 +107,12 @@ public HighlightField highlight(HighlighterContext highlighterContext) {
             textsToHighlight = HighlightUtils.loadFieldValues(field, mapper, context, hitContext);
 
             for (Object textToHighlight : textsToHighlight) {
-                String text = textToHighlight.toString();
+                String text;
+                if (textToHighlight instanceof BytesRef) {
+                    text = mapper.fieldType().valueForDisplay(textToHighlight).toString();
+                } else {
+                    text = textToHighlight.toString();
+                }
```

Comment:
Should we just do `String text = mapper.fieldType().valueForDisplay(textToHighlight).toString();` all the time? This looks correct to me and is more future-proof? If that does not work, feel free to merge the PR as-is.